RPC v0.8.0

thiagodeev opened the pull request:
Issue: https://github.com/NethermindEth/starknet.go/issues/623

This PR aggregates all v0.8.0 changes

thiagodeev committed:
Thiagodeev/rpcv08 write methods, blockHeader and error 53 (#626)

Update write methods, error 53 and block header for rpc08

thiagodeev committed:
Adjusts TransactionStatus

thiagodeev committed:
Adjusts EstimateFee

thiagodeev committed:
Fixes blockTest error

thiagodeev committed:
Merge branch 'main' into v0.8.0

thiagodeev committed:
Merge branch 'main' into v0.8.0

thiagodeev committed:
Adds GetStorageProof method (#635)

* implement rpcv8 GetStorageProof method

thiagodeev committed:
Adds GetMessagesStatus method (#634)

* Added GetMessagesStatus method

* PR comment fixes

thiagodeev committed:
starknet_getCompiledCasm (#642)

* implement starknet_getCompiledCasm for RPCv8

thiagodeev committed:
Merge branch 'main' into v0.8.0

thiagodeev committed:
Thiagodeev/rpcv08 final changes (#644)

* Adjusts TransactionStatus

* Adjusts EstimateFee

* Fixes blockTest error

* Adds GetStorageProof method (#635)

* implement rpcv8 GetStorageProof method

* Adds GetMessagesStatus method (#634)

* Added GetMessagesStatus method

* PR comment fixes

* Fix MerkleNode type

* Updates PendingBlockReader type

* Updates ExecutionResources and remove ComputationResources

* Fixes failing tests and improves Error() resp

* Adds TODO comments and set a pointer to Marshal method

* Thiagodeev/rpcv08 write methods, blockHeader and error 53 (#626)

Update write methods, error 53 and block header for rpc08


* Adjusts EstimateFee

thiagodeev committed:
New rpcdata + RPCv08 errors (#649)

Update RPC rrror structure to handle RPv0.8.0 errors

thiagodeev committed:
Merge branch 'main' into v0.8.0

thiagodeev committed:
Fixes failing test

rianhughes commented on rpc/transaction_test.go:
```diff
@@ -179,6 +179,8 @@ func TestTransactionReceipt(t *testing.T) {
 
 // TestGetTransactionStatus tests starknet_getTransactionStatus
 func TestGetTransactionStatus(t *testing.T) {
+	//TODO: implement a test case to 'failure_reason' before merge
```
t.Run("TODO: implement a test case to 'failure_reason' before merge", func(t *testing.T) {
		t.Skip("TODO: implement a test case to 'failure_reason' before merge")
	})

rianhughes commented on rpc/types_contract.go:
```diff
@@ -69,11 +69,87 @@ type ContractClass struct {
+	ClassesProof           NodeHashToNode   `json:"classes_proof"`
+	ContractsProof         ContractsProof   `json:"contracts_proof"`
+	ContractsStorageProofs []NodeHashToNode `json:"contracts_storage_proofs"`
+	GlobalRoots            []NodeHashToNode `json:"global_roots"`
```
Is this correct? Spec indicates it should be
```
"global_roots": {
              "type": "object",
              "properties": {
                "contracts_tree_root": {
                  "$ref": "#/components/schemas/FELT"
                },
                "classes_tree_root": {
                  "$ref": "#/components/schemas/FELT"
                },
                "block_hash": {
                  "description": "the associated block hash (needed in case the caller used a block tag for the block_id parameter)",
                  "$ref": "#/components/schemas/FELT"
                }
              },
              "required": ["contracts_tree_root", "classes_tree_root", "block_hash"]
            }
```

rianhughes replied:
I guess it's meant to be `GlobalRoots`

rianhughes replied:
This has been updated 
```
type StorageProofResult struct {
...
	GlobalRoots            []GlobalRoots    `json:"global_roots"`
}
```

rianhughes commented on rpc/errors.go:
```diff
@@ -194,4 +272,74 @@ var (
+	temp = contractEx.Message
+
+	return json.Marshal(temp)
+}
```
suggestion

```
func (contractEx *ContractExecutionError) MarshalJSON() ([]byte, error) {
	if contractEx.ContractExecutionErrorInner != nil {
		return json.Marshal(contractEx.ContractExecutionErrorInner)
	}
	return json.Marshal(contractEx.Message)
}
```

rianhughes replied:
Outdated

rianhughes commented on rpc/errors.go:
```diff
@@ -62,19 +65,90 @@ func tryUnwrapToRPCErr(err error, rpcErrors ...*RPCError) *RPCError {
+	temp = rpcData.Message
+
+	return json.Marshal(temp)
 }
```
suggestion
```
func (rpcData *RPCData) MarshalJSON() ([]byte, error) {
	if rpcData.CompilationErrorData != nil {
		return json.Marshal(rpcData.CompilationErrorData)
	}

	if rpcData.ContractErrorData != nil {
		return json.Marshal(rpcData.ContractErrorData)
	}

	if rpcData.TransactionExecutionErrorData != nil {
		return json.Marshal(rpcData.TransactionExecutionErrorData)
	}
	return json.Marshal(rpcData.Message)
}
```

rianhughes replied:
Outdated

thiagodeev committed:
Merge branch 'main' into v0.8.0

thiagodeev committed:
Merge branch 'main' into v0.8.0

thiagodeev committed:
thiagodeev/rpcv08-websocket (#651)

* Update go.mod to include gorilla/websocket and enhance RPC provider with Websocket support

- Added gorilla/websocket v1.5.3 as a direct dependency in go.mod.
- Introduced NewWebsocketProvider function in provider.go to create a Websocket RPC Provider instance, enhancing the existing HTTP provider functionality.

* Enhance NewWebsocketProvider with cookie support in provider.go

* Added new errors

* Refactor RPC client and enhance WebSocket support

- Removed the dependency on github.com/ethereum/go-ethereum and replaced it with github.com/NethermindEth/starknet.go/client in go.mod and related files.
- Introduced new WebSocket provider functionality in provider.go

* adds remaining test data and improves tests, and adapt subscribe method to work on starknet

* Enhance Starknet WebSocket support and subscription handling

- Updated the handler to correctly process subscription IDs for Starknet, accommodating the new structure returned by the Starknet API.
- Modified the subscriptionResult struct to include a Starknet-specific subscription ID field.
- Adjusted the WebSocket provider to support new block header subscriptions, improving the overall subscription mechanism.
- Updated example usage to reflect changes in subscription handling and error management.

These changes improve compatibility with Starknet's API and enhance the robustness of the WebSocket client.

* improves newHeads method and WS close message

* Enhance Starknet compatibility and testing framework

- Updated the main CI workflow to include testing for the client with mocks.
- Modified subscription handling in `subscription.js` to accommodate the new `subscription_id` structure from Starknet.
- Refactored `provider_test.go` to include WebSocket provider support and improved test configurations.
- Introduced a new test file `websocket_test.go` to validate WebSocket subscriptions for new block headers, ensuring robust error handling and compatibility with the testnet environment.

These changes improve the overall robustness of the RPC client and enhance compatibility with Starknet's API.

* Fixes linter and test errors

* draft implementation of subscribeEvents

* client adaptation and functional version of subscribeEvents

Note: at the moment it's not possible to omit optional parameters in json-rpc calls using array as structure type with Juno, and the current client implementation only supports sending parameters as arrays. Therefore, I changed the Subscribe function and now we are able to send optional parameters as object too. That way Juno doesn't return an error

* Implement SubscribeEvents functionality in WebSocket provider with enhanced error handling and testing

- Added the SubscribeEvents method to the WebSocket provider, allowing for event subscriptions with optional parameters.
- Introduced a new test case in websocket_test.go to validate event subscriptions under various conditions, including handling of empty arguments and specific address filtering.
- Updated the HexToFeltNoErr utility function to convert hexadecimal strings to felt objects without error handling, suitable for internal use.

These changes improve the robustness of the WebSocket client and enhance compatibility with Starknet's event subscription model.

* finish subscribeEvents tests

* add starknet-spec updates

* Finish subscribeTransactionStatus

* Implement SubscribePendingTransactions

* revert mistake in the account_test.go file

* improve SubscribeNewHeads param

* improve SubscribeEvents method

* Add SubscribePendingTransactions tests

* Add support for Reorg notifications

* add TestUnsubscribe

* Create WebsocketProvider interface

* Remove some TODO comments

* Reorg ws methods alphabetically

* Improve ws methods description

* Add WebSocket example with detailed usage and README update

* Update README with WebSocket example documentation

* Update CI workflow to build WebSocket and typedData projects

* Remove HexToFeltNoErr

* Refactor WebSocket subscription types with SubscriptionBlockID and remove unused error

* Fix PR review comments

* Improve TestSubscribeEvents testcase description and fix PR comment

* Fix SubscribeTransactionStatus with new Juno update

* New 'is_reverted' field from spec update

* Update 'subscription_id' handling

* Compare error with the expected one in TestSubscribePendingTransactions

* Add TestReorgEvents placeholder

* Replace INTEGRATION_BASE to HTTP_PROVIDER_URL, and add WS_PROVIDER_URL placeholder

Also, change 'beforeEach' func to also initialize a websocket provider

* change 'beforeEach' func to validate ws tests

* fix ws tests error in 'devnet' env

* Fix ws tests when running in other envs

* Fix TestSubscribePendingTransactions test and types

* Rename env in github actions

thiagodeev committed:
Thiagodeev/rpcv08 - implement/fix/update remaining tests for rpcv08 (#662)

* New 'is_reverted' field

* TestGetTransactionStatus updated to test new "FailureReason" field

* Fix bug in account test and improves setup.go log in panic

* Update TestEstimateMessageFee test

* Update TestEstimateFee test

* Draft implementation of TestGetStorageProof test

* Create UnmarshallFileToType func to unmarshal json to types

* Push 'compiledCasm.json' file to repo

* Draft TestCompiledCasm test

* Fix errors, and improve Unmarshal and validations in the executable types

* Improves Unmarshal and Marshal of executable types

* Refactor types_executables.go types

* TestCompiledCasm passing testnet test case

* Add ErrClassHashNotFound test case to TestCompiledCasm

* Add mock tests for TestCompiledCasm

* TestEstimateMessageFee passing

* Fix TestEstimateFee

* Fix TestSimulateTransaction

* Fix TestBlockWithReceipts

* FixTestBlockWithTxHashes

* Fix TestBlockWithTxs testnet

* Update TestSpecVersion test

* Change tests to use the new UnmarshallFileToType func

* Rename UnmarshallFileToType to TestUnmarshallFileToType

* New UnmarshallFileToType utility

* Rename/improve UnmarshallFileToType and adapt code to use it

* Add TestGetMessagesStatus test

* Fix mainnet tests

* Fix merge conflicts

* Remove 'break' from loops

* Rename UnmarshallJSONFileToType func and improves TestUnmarshallJSONFileToType

* Create TraceStatusData type

* FIx TestGetTransactionStatus and TestGetMessagesStatus tests

* Refactor RPCData

* improvements in contract_test.go

* Fix post merge conflits

* make ErrorMessage() return the entire ContractExecutionError data

thiagodeev committed:
Thiagodeev/rpcv08 refactor tests and utils organization (#672)

* refactor the tests organization

* reorg utils files to avoid import circle

* refactor devnet pkg tests

* rename import name in utils pkg

* Refactor import names in multiple packages

Update import statements to use `internalUtils` instead of `utils` across various files to resolve import circle and improve package organization. This change affects multiple packages including account, contracts, curve, devnet, hash, rpc, typedData, and utils.

* Fix account_test.go bug of not running mock tests

* rename Felt.go utils to lower case

* New testcase for TestCall

* Update TestSimulateTransaction test data

* Fix TestSubscribeEvents

* Skip failing tests due to bugs or outdates

* Update devnet to 0.2.4 and fix TestStorageAt on it

thiagodeev committed:
Thiagodeev/rpcv08 new txn utilities (#673)

* Create utility for build transactions

* Create BuildInvokeTxn account method

* New InvokeFunctionCall type and InvokeFuncCallsToFunctionCalls util

* create FeeEstimationToResourceBoundsMapping util

* Set default value of U64 and U128 to '0x0'

* Add MarshalJSON method and validation for DataAvailabilityMode type

* create BuildAndSendInvokeTxn

* Revert "Set default value of U64 and U128 to '0x0'"

This reverts commit 33e92dfbefb097174ef5eecaefb9f8e300173f8f.

* Update Resource Bounds hash calculation to include the new field

* New multiplier param for BuildAndSendInvokeTxn func

* New tests placeholders

* Make BuildDeclareTxn accept casmClass instead of compiledClassHash

* Update TransactionHashDeclare to accept BroadcastDeclareV3Txn

* New BuildAndSendDeclareTxn method

* new draft simpleDeclare example for testing the BuildAndSendDeclareTxn impl

* New BuildAndSendDeployAccountTxn method

* Turn PrecomputeAccountAddress method from account type into a normal func

* New ResBoundsMapToOverallFee and fee utils, also tests for them

* Change BuildAndSendDeployAccountTxn to BuildAndEstimateDeployAccountTxn

* Update deployUDC example just to pass CI, needs to be updated in future

* Update TestTransactionHashDeclare, TransactionHashDeclare now validates DeclareTxnV3

* Update TestTransactionHashInvokeV3

* Update TestTransactionHashdeployAccount

* Update TestSendInvokeTxn test

* Update TestSendDeclareTxn

* Update TestSendDeployAccountDevnet

* Add skiped test description

* Improve BuildAndEstimateDeployAccountTxn desciption to include information about funding preComputedAdd

* Add makeResourceBoundsMapWithZeroValues helper

* Fix typo in utils/transaction.go

* Improves and test FeeEstToResBoundsMap func

* Improve and test ResBoundsMapToOverallFee

* make WeiToETH and ETHToWei use felts instead of big.int

* Remove some testcases and improve func description

* Handle error in simpleInvoke example

* Helper internal func to ResBoundsMapToOverallFee test

* Update TestBuildAndSendInvokeTxn test

* Remove unused secret and add new secrets to account tests in CI

* Implement TestBuildAndSendDeclareTxn test

* Implement TestBuildAndSendDeclareTxn test

thiagodeev committed:
Thiagodeev/rpcv08 update examples (#677)

* Rename simpleInvoke example and star to modify it

* New acc.Provider() method

* Fix TransactionReceiptWithBlockInfo unmarshal error

* Finish descriptiveInvoke func

* Rename descriptiveInvoke to verboseInvoke

* Finish simpleInvoke

* Finish main

* Finish invoke test, update readmes

* Finish deployAccount example

* Finish deployContractUDC

* Finish simpleCall example

* Improving simpleDeclare example

* Renaming files, improving readmes

* Update readme. Finish simpleDeclare example

* Typo. Rename UnmarshallJSONFileToType UnmarshalJSONFileToType

* Final typos and README adjustments

* Fix typos and README

* include simpleDeclare build on CI

thiagodeev committed:
Thiagodeev/rpcv08 final adjustments (#678)

* Update README for new release

* Fix optional field being required by TxnTrace types

* Fix TxnTrace field errors

* Fix TestTransactionTrace on testnet

* Improve test.go logs

* Remove the SimulateTransactionOutput type

* Fix and improve TestTransactionTrace and TestSimulateTransaction tests

* Update TestTraceBlockTransactions testdata

* Improve TestTraceBlockTransactions coverage

* Fix wrong field name in the InitSquashData type

* Fix TestUnmarshalCasmClassHash, started to refactor the contract types

* Rename contract.CasmCompiledContractClass to CasmClass

* Move Sierra types from rpc pgk to contracts

* Update imports and rename files

* Fix TestTransactionTrace and TestTraceBlockTransactions on mock env

* Fix TestCompiledCasm test

* Fix TestBuildAndSendDeclareTxn import error

* Fix linter err

* Improve BuildAndSend account tests to use WaitForTransactionReceipt method

* Update TestCompiledCasm test with Juno's fix

* Undo TestCall skip

* Small spec updates

thiagodeev committed:
Thiagodeev/rpcv08 final adjustments v2 (#680)

* Create CallContextWithSliceArgs client method

* Start to improve GetStorageProof method and tests

* Fix MerkleNode type

* Implement TestGetStorageProof test

* Fix TestClientSubscriptionUnsubscribeServer test

* Fix TestClientNotificationStorm test

* Small TODO fixes

* Remove mainnet testcase from TestSimulateTransaction

* Update TestBlockWithReceipts mainnet test data

* Fix mainnet tests on websocket

* Allow CI to run testnet and mainnet tests

* Make struct values be passed by pointer in some functions

* Make err.Error() return all RPCError fields

* Remove PanicRPC func since now RPCErrors return all errors in err.Error() method

* Update TestRPCError test

* Update TestGetStorageProof test with Juno fixes

* Update TestTransactionByBlockIdAndIndex test data

* Fix TestEstimateFee test

* Fix error message in tests

* Make SignInvokeTransaction available to all invoke types, and new func to calc invoke hashes for each invoke version

* Make SignDeployAccountTransaction available to all deployAccTxn types, and new func to calc deployAcc hashes for each deployAcc version

* Make SignDeclareTransaction available to all declare types, and new func to calc declare hashes for each declare version

* Remove duplicated Provider methods from Account type

* Remove SKIP_VALIDATE flag from Build helpers

* Call the new Nonce method

thiagodeev committed:
Thiagodeev/rpcv08 final adjustments v3 (#681)

* Remove deprecated broadcast txns types

* Fix TestDeclareTransaction on mock

* Fix other mock failing tests

* Fix TestGetStorageProof random fails

* Fix TestBuildAndSendDeclareTxn test

* Doc fixes

* Create func comments in the acc pkg

* Generate account moks

* Move txn hash functions from the account pkg to the hash pkg

* New HexToU256Felt util

* New U256FeltToHex util

* New fillEmptyFeeEstimation temp helper to zero fee values from estimation

* Change code to use the new HexToU256Felt method

* Make DataAvailabilityModeConc and TipAndResourcesHash public

* Remove BlockIDWithoutPending and SubscriptionBlockID types. Adapt the BlockID to fit them

* Fix TestSubscribeEvents test

* Improve TestSubscribeEvents keys tescase
